### ISSUE TITLE
404 when trying to edit or remove path containing a slash

### ISSUE THREAD

**Issue by squareclouds** [distantnative/retour-for-kirby#300]
hi everyone,

i just realized that when i have a path with a slash,  lets say 

blog-referenzen/(:all)

or

de/uebermich

that i am redirecting, it works fine, but when i try to either remove or edit them in the panel the popup/modal does not open and i get redirected to 404, so i have to delete them by hand in the yaml

--

## Steps to reproduce

1. setup a path with a slash to be redirected somewhere
2. after saving it, try to edit it

## Expected behavior

modal opens up and you can edit (or delete) the entry

## Additional context

![routes-error](https://user-images.githubusercontent.com/9073054/160383029-3c6c35b6-0a3a-42d0-a0f2-a3542cc208c1.gif)

## Context

- Browser: latest FF and Chrome
- Kirby version: 3.6.2
- Retour version: 4.0.0

**Comment by medienbaecker** [comment]
This also happens when you try to resolve or delete entries in the "failures" tab.

**Comment by createdbyarc** [comment]
Hi also experiencing something similar - unfortunately I need to redirect a lot of urls with forward slashes - thanks in advance.

![retour](https://user-images.githubusercontent.com/58830123/169522442-16027f2b-f256-440c-ac3b-e41b8b4a46e3.gif)

**Comment by bogdancondorachi** [comment]
> Hi also experiencing something similar - unfortunately I need to redirect a lot of urls with forward slashes - thanks in advance.
> 
> ![retour](https://user-images.githubusercontent.com/58830123/169522442-16027f2b-f256-440c-ac3b-e41b8b4a46e3.gif) [ ](https://user-images.githubusercontent.com/58830123/169522442-16027f2b-f256-440c-ac3b-e41b8b4a46e3.gif) ![retour](https://user-images.githubusercontent.com/58830123/169522442-16027f2b-f256-440c-ac3b-e41b8b4a46e3.gif) [ ](https://user-images.githubusercontent.com/58830123/169522442-16027f2b-f256-440c-ac3b-e41b8b4a46e3.gif)

Happens to me as well when trying to edit or delete an redirect

**Comment by distantnative** [comment]
I am sorry that this hasn't been fixed. I keep trying to reproduce the bug on my side from your screencasts etc but no luck, always works fine for me.

Can you all share the following details:
- what's the path of that entry
- what's the redirect to of that entry
- where does the Panel take you when you try to edit (which URL)

Thanks a lot!

**Comment by bogdancondorachi** [comment]
Hi @distantnative, thanks for getting back. I know you are busy on Kirby right now. Here's my details on the bug

These are my entries
![image](https://user-images.githubusercontent.com/39380741/175773551-6ee2d2d0-8a73-434f-beb3-492d8e1c3f1e.png)

and if I try to edit them or delete them, I get a 404
the url seems looks like this "https://hasty.ai/panel/dialogs/retour/redirects/content-hub%2Fsoftware-engineer/edit"
![image](https://user-images.githubusercontent.com/39380741/175773635-65f9b54a-58f1-4a87-8ce6-49107c6ad620.png)

Also, I get the same 404 on any failures entry when I try to 'add as redirect' or 'remove' an entry (clear log works fun)

My kirby runs on 3.6.6 on php 8.0.20 and apache

**Comment by code-blep** [comment]
I have also encountered this issue on a clients website. **This issue is easily replicated by installing the demo version of Kirby.** If the URL (after the TLD stuff) has a slash in it, it breaks. Hate to say it but this makes this plugin pretty much unusable. 

After various tests, I suspect that the issue is in the index.js file, but because it has been minified, it is not practical to work on.

**Comment by FynnZW** [comment]
Having the same issue. Any slash breaks opening the edit modal. 
The url is `{{siteurl}}/panel/dialogs/retour/redirects/projekte%2Ftest/edit`

**Comment by tobimori** [comment]
Might this be an Apache issue? It runs fine on the PHP dev server, and my nginx instances, but not on production Apache.

**Comment by ieteo** [comment]
1.
Adding `AllowEncodedSlashes NoDecode` fixed the issue on Apache instances. Thanks!

2. 
> In order to fix this issue, no matter the servers' settings, it would be beneficial to switch to another way of encoding the path

Sharing the same thought
+1

**Comment by distantnative** [comment]
@ieteo thanks for the setup fix for everyone and also thanks for figuring out why - I was really clueless since it worked in all my setups. I will have a look how we could encode the path differently.

**Comment by ieteo** [comment]
@distantnative the credit goes to @tobimori. He is the one who figured it out and proposed possible fix in code regardless of the web server used. 

His idea to base64 encode the path is quite interesting. I have just tested the encoding between JavaScript (FF and Edge) and PHP8 and it works like charm.

**JS**
`btoa("/path/subpath/etc/document?param1=foo&param3=bar")`
outputs:
`L3BhdGgvc3VicGF0aC9ldGMvZG9jdW1lbnQ/cGFyYW0xPWZvbyZwYXJhbTM9YmFy"`

**PHP**
`\base64_decode('L3BhdGgvc3VicGF0aC9ldGMvZG9jdW1lbnQ/cGFyYW0xPWZvbyZwYXJhbTM9YmFy')`
outputs:
`/path/subpath/etc/document?param1=foo&param3=bar`


and it can also be done the other way around

**PHP** 
`\base64_encode('/path/subpath/etc/document?param1=foo&param3=bar')`
outputs:
`L3BhdGgvc3VicGF0aC9ldGMvZG9jdW1lbnQ/cGFyYW0xPWZvbyZwYXJhbTM9YmFy`

**JS**
`atob("L3BhdGgvc3VicGF0aC9ldGMvZG9jdW1lbnQ/cGFyYW0xPWZvbyZwYXJhbTM9YmFy")`
outputs:
`/path/subpath/etc/document?param1=foo&param3=bar`


To summarize

PHP **base64_encode()** is equivalent to JS **btoa()**
PHP **base64_decode()** is equivalent to JS **atob()**

**Comment by tobimori** [comment]
> @distantnative the credit goes to @tobimori. He is the one who figured it out and proposed possible fix in code regardless of the web server used.
> 
> His idea to base64 encode the path is quite interesting. I have just tested the encoding between JavaScript (FF and Edge) and PHP8 and it works like charm.
> 
> **JS** `btoa("/path/subpath/etc/document?param1=foo&param3=bar")` outputs: `L3BhdGgvc3VicGF0aC9ldGMvZG9jdW1lbnQ/cGFyYW0xPWZvbyZwYXJhbTM9YmFy"`
> 
> **PHP** `\base64_decode('L3BhdGgvc3VicGF0aC9ldGMvZG9jdW1lbnQ/cGFyYW0xPWZvbyZwYXJhbTM9YmFy')` outputs: `/path/subpath/etc/document?param1=foo&param3=bar`
> 
> and it can also be done the other way around
> 
> **PHP** `\base64_encode('/path/subpath/etc/document?param1=foo&param3=bar')` outputs: `L3BhdGgvc3VicGF0aC9ldGMvZG9jdW1lbnQ/cGFyYW0xPWZvbyZwYXJhbTM9YmFy`
> 
> **JS** `atob("L3BhdGgvc3VicGF0aC9ldGMvZG9jdW1lbnQ/cGFyYW0xPWZvbyZwYXJhbTM9YmFy")` outputs: `/path/subpath/etc/document?param1=foo&param3=bar`
> 
> To summarize
> 
> PHP **base64_encode()** is equivalent to JS **btoa()** PHP **base64_decode()** is equivalent to JS **atob()**

Just a hint: Base64 contains slashes by itself, so we would need to replace them after encoding/before encoding manually to make it "URL-safe".
